### PR TITLE
Add grid component

### LINKED ISSUE #1154: Add Grid component

### PR DESCRIPTION
**Pull request by tiny-dust**
## Checklist

List of tasks you have already done and plan to do.

- [x] Fix linting errors
- [x] Tests have been added / updated (or snapshots)

## Change information

The grid component provides a new layout and supports the following parameters.
#### Grid Props

| Prop | Description | Type             | Default       |
| ------- | --- |----------------|-----------|
| `column`    | Set Column Num  | _string \| number_ | `4` |
| `iconSize`  | Set Icon Size | _string \| number_ | `24` |
| `square`    | Set to a square, it is recommended that the number of columns be less than `4` | _boolean_ | `false` |
| `gutter`    | Set Gutter | _string \| number_ | `0` |
| `center`    | Center the content | _boolean_ | `false` |
| `border`    | Set Border | _boolean_ | `false` |
| `ripple`    | Set click water ripple effect | _boolean_ | `false` |
| `direction` | Arrangement direction of the grid content, the optional value is`horizontal`、`vertical` | _string_ | `vertical` |

#### GridItem Props

| Prop | Description | Type             | Default       |
| ------- | --- |----------------|-----------|
| `text`  | Text | _string_ | `` |
| `icon`  | Icon | _string_ | `` |
| `iconColor`  | Set Icon color | _string_ | `` |
| `badge`  | Set Badge value | _string \| number_ | `` |
| `dot`  | Set Dot | _boolean_ | `false` |
| `badgeProps`  | Badge Configuration | _BadgeProps_ | `` |

## Issues

close #1154 .

### CODE REVIEW

File: packages/varlet-ui/src/grid/props.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    default: 0,
+    validator(value: string | number) {
+      const val = toNumber(value)
+      return val >= 0 && val <= 24
+    },
+  },
+  direction: {
+    type: String as PropType<'horizontal' | 'vertical'>,
+    default: 'vertical',
+    validator(value: string) {
```
Comment (chouchouji) [inline packages/varlet-ui/src/grid/props.ts:48]: We had removed all validators and simplified types in props.ts.  You can find all changes in recent commits.
image

Comment (tiny-dust) [reply packages/varlet-ui/src/grid/props.ts]: ok~ I'm done

File: packages/varlet-ui/src/grid-item/props.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  },
+  iconColor: {
+    type: String,
+    default: '',
+  },
+  badge: {
+    type: [String, Number],
+    default: '',
+  },
+  dot: {
```
Comment (chouchouji) [inline packages/varlet-ui/src/grid-item/props.ts:22]: This prop can be simplified to `dot: Boolean`.

File: packages/varlet-ui/src/grid-item/props.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    default: '',
+  },
+  iconColor: {
+    type: String,
+    default: '',
+  },
+  badge: {
+    type: [String, Number],
+    default: '',
+  },
```
Comment (haoziqaq) [inline packages/varlet-ui/src/grid-item/props.ts:21]: These props should not specify a default value: '', but should be undefined

File: packages/varlet-ui/src/grid-item/GridItem.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      const style: StyleValue = {
+        flex: `0 calc(${size}% - ${gutter.value}px)`,
+      }
+
+      if (gutter && gutter.value > 0) {
+        style.margin = `${gutter.value / 2}px`
+      }
+
+      return style
+    })
```
Comment (haoziqaq) [inline packages/varlet-ui/src/grid-item/GridItem.tsx:65]: replace computed with inline function

File: packages/varlet-ui/src/grid-item/GridItem.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    const elevationCls = computed(() => {
+      if (elevation.value || gutter.value > 0) {
+        return `var-elevation--${elevation.value || 2}`
+      }
+      return ''
+    })
```
Comment (haoziqaq) [inline packages/varlet-ui/src/grid-item/GridItem.tsx:72]: replace this with classes function

File: packages/varlet-ui/src/grid-item/GridItem.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    const gridItemProvide: GridItemProvide = {
```
Comment (haoziqaq) [inline packages/varlet-ui/src/grid-item/GridItem.tsx:74]: typo: provider

File: packages/varlet-ui/src/grid-item/GridItem.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      () => {
+        if (square.value && gridElement.value) {
+          const width = gridElement.value.offsetWidth
+          gridElement.value.style.height = `${width}px`
+        }
+      },
+      {
+        immediate: true,
+      }
+    )
```
Comment (haoziqaq) [inline packages/varlet-ui/src/grid-item/GridItem.tsx:90]: extract a function and replace immediate: true with onSmartMounted

File: packages/varlet-ui/src/grid/Grid.tsx [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+export default defineComponent({
+  name: 'VarGrid',
+  props,
+  setup(props, { slots }) {
+    const { bindItem } = useGridItem()
+
+    const gutter = computed(() => toPxNum(props.gutter))
+
+    const gridProvide: GridProvider = {
```
Comment (haoziqaq) [inline packages/varlet-ui/src/grid/Grid.tsx:20]: type: provider

File: packages/varlet-ui/src/grid/Grid.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      elevation: computed(() => toNumber(props.elevation) || 0),
+    }
+    bindItem(gridProvide)
+
+    const customStyle = computed(() => {
+      const style = {
+        margin: `-${gutter.value / 2}px`,
+      }
+      return style
+    })
```
Comment (haoziqaq) [inline packages/varlet-ui/src/grid/Grid.tsx:38]: Unnecessary, replace it with inline style.

**Comment by haoziqaq** [comment]
GridItem Props `badge` `dot` `badgeProps` can be merge. ref to [BottomNavigation](https://varlet.gitee.io/varlet-ui/#/en-US/bottom-navigation#Show%20Badge).

Square Gird Items should resize on window resize.

File: packages/varlet-ui/src/grid/__tests__/index.spec.js [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        </div>
+      `,
+    }
+
+    const wrapper = mount(Wrapper)
+
+    await nextTick()
+    const gridItems = wrapper.find('.var-grid-item')
+
+    expect(gridItems.classes()).toContain('var-grid-item__square')
```
Comment (songjianet) [inline packages/varlet-ui/src/grid/__tests__/index.spec.js:108]: Please add unmount.

**Review by songjianet — APPROVED** [review]

File: packages/varlet-ui/src/grid/__tests__/index.spec.js (hunk 10 above)
Comment (songjianet) [inline packages/varlet-ui/src/grid/__tests__/index.spec.js:108]: Please add unmount.

File: packages/varlet-ui/src/grid/__tests__/index.spec.js [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+  })
+
+  test('test grid gutter', () => {
+    const wrapper = mount(Wrapper, {
+      props: {
+        gutter: '10px',
+      },
+    })
+    expect(wrapper.find('.var-grid-item').attributes('style')).toContain('margin: 5px')
```
Comment (songjianet) [inline packages/varlet-ui/src/grid/__tests__/index.spec.js:117]: Please add unmount.

File: packages/varlet-ui/src/grid/__tests__/index.spec.js [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    })
+    const prefix = 'var-grid-item'
+
+    expect(wrapper.find(`.${prefix}`).classes()).not.toContain(`${prefix}__border`)
+    expect(wrapper.find(`.${prefix}`).classes()).not.toContain(`${prefix}__center`)
+
+    wrapper.setProps({ border: true, center: true })
+    await delay(100)
+    expect(wrapper.find(`.${prefix}`).classes()).toContain(`${prefix}__border`)
+    expect(wrapper.find(`.${prefix}`).classes()).toContain(`${prefix}__center`)
```
Comment (songjianet) [inline packages/varlet-ui/src/grid/__tests__/index.spec.js:135]: Please add unmount.

File: packages/varlet-ui/src/grid/__tests__/index.spec.js [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+    const wrapper = mount(Wrapper)
+    const gridItems = wrapper.findAll('.var-grid-item')
+
+    expect(gridItems[0].find('.var-badge--dot').exists()).toBe(true)
+    expect(gridItems[1].find('.var-badge--dot').exists()).toBe(false)
+    expect(gridItems[1].find('.var-badge__value').element.innerHTML).toBe('66')
+    expect(gridItems[2].find('.var-badge--primary').exists()).toBe(true)
+    expect(gridItems[3].find('.var-badge--right-bottom').exists()).toBe(true)
+    expect(gridItems[4].find('.var-badge').exists()).toBe(false)
```
Comment (songjianet) [inline packages/varlet-ui/src/grid/__tests__/index.spec.js:163]: Please add unmount.

File: packages/varlet-ui/src/grid/__tests__/index.spec.js [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+  }
+  const wrapper = mount(Wrapper)
+  const gridItem = wrapper.find('.var-grid-item')
+
+  expect(gridItem.find('.icon').exists()).toBe(true)
+  expect(gridItem.find('.icon').text()).toBe('icon')
+  expect(gridItem.find('.text').exists()).toBe(true)
+  expect(gridItem.find('.text').text()).toBe('text')
+  expect(gridItem.find('.default').exists()).toBe(true)
+  expect(gridItem.find('.default').text()).toBe('default')
```
Comment (songjianet) [inline packages/varlet-ui/src/grid/__tests__/index.spec.js:228]: Please add unmount.

**Review by songjianet — CHANGES_REQUESTED** [review]

File: packages/varlet-ui/src/grid/__tests__/index.spec.js (hunk 14 above)
Comment (tiny-dust) [reply packages/varlet-ui/src/grid/__tests__/index.spec.js:228]: ok~

**Pull request closed without merging**